oauth_provider/controllers/main.py

```diff
@@ -0,0 +1,316 @@
+        """
+        token = http.request.env['oauth.provider.token'].search([
+            ('token', '=', access_token),
+        ])
+        if not token:
+            return json.dumps({'error': 'invalid_or_expired_token'})
+        oauth2_server = token.client_id.get_oauth2_server()
+
+        # Retrieve needed arguments for oauthlib methods
+        uri = http.request.httprequest.base_url
+        http_method = http.request.httprequest.method
+        body = oauthlib.common.urlencode(
+            http.request.httprequest.values.items())
+        headers = http.request.httprequest.headers
+
+        valid, oauthlib_request = oauth2_server.verify_request(
+            uri, http_method=http_method, body=body, headers=headers)
+        if not valid:
+            return json.dumps({})
+
+        token_lifetime = (fields.Datetime.from_string(token.expires_at) -
+                          datetime.now()).seconds
+        # Base data to return
+        data = {
+            'audience': token.client_id.identifier,
+            'scopes': ' '.join(token.scope_ids.mapped('code')),
+            'expires_in': token_lifetime,
+        }
+
+        # Add the oauth user identifier, if user's information access is
+        # allowed by the token's scopes
+        user_data = token.get_data_for_model(
+            'res.users', res_id=token.user_id.id)
+        if 'id' in user_data:
+            # Include the client and user identifiers in the final identifier
+            # to generate a different identifier for the same user, depending
+            # on the client accessing this user
+            # By doing this, clients cannot find a list of common users by
+            # comparing their identifiers list from tokeninfo
+            user_identifier = token.client_id.identifier
+            user_identifier += token.user_id.oauth_identifier
+            # Use a sha256 to avoid a too long final string
+            data.update(user_id=hashlib.sha256(user_identifier).hexdigest())
+        return json.dumps(data)
+
+    @http.route('/oauth2/userinfo', type='http', auth='none', methods=['GET'])
+    def userinfo(self, access_token=None, *args, **kwargs):
+        """
+        Return some information about the user linked to the supplied token
+        Similar to Google's "userinfo" request
+        """
+        token = http.request.env['oauth.provider.token'].search([
+            ('token', '=', access_token),
+        ])
+        if not token:
+            return json.dumps({'error': 'invalid_or_expired_token'})
+        oauth2_server = token.client_id.get_oauth2_server()
+
+        # Retrieve needed arguments for oauthlib methods
+        uri = http.request.httprequest.base_url
+        http_method = http.request.httprequest.method
+        body = oauthlib.common.urlencode(
+            http.request.httprequest.values.items())
+        headers = http.request.httprequest.headers
+
+        valid, oauthlib_request = oauth2_server.verify_request(
```

Comment:
This token & argument retrieval followed verification block seems to be repeated a lot. It would be nice if it was just a helper method instead.